function/function.rs

```diff
@@ -4,95 +4,27 @@
  * file, You can obtain one at https://mozilla.org/MPL/2.0/.
  */
 
-use answer::Type;
-use encoding::{graph::definition::definition_key::DefinitionKey, value::value_type::ValueType};
-use ir::{
-    pattern::variable_category::{VariableCategory, VariableOptionality},
-    program::function::FunctionIR,
-};
+use encoding::graph::definition::{definition_key::DefinitionKey, function::FunctionDefinition};
+use ir::program::function_headers::FunctionIDTrait;
 
-/// Function represents the user-defined structure:
-/// fun <name>(<args>) -> <return type> { <body> }
-pub struct Function {
-    definition_key: DefinitionKey<'static>,
+use crate::FunctionManagerError;
 
-    // parsed representation
-    name: String,
-    arguments: Vec<FunctionArgument>,
-    return_type: FunctionReturn,
-
-    // pre-compiled arguments, body, return // TODO maybe deferred compilation
-    ir_body: FunctionIR,
-}
-
-struct FunctionArgument {
-    name: String,
-    type_: FunctionValuePrototype,
-}
-
-enum FunctionReturn {
-    Stream(Vec<FunctionValuePrototype>),
-    Single(FunctionValuePrototype),
-}
-
-impl Function {
-    // TODO: receive a string, which can either come from the User or from Storage (deserialised)
-    //       will require type manager to convert labels into Types and Definitions etc.
-    fn new(definition: &str) -> Self {
-        // 1. parse into TypeQL
-        // 2. extract into data structures
-        // possible later: // TODO: what if recursive - the function call will need information from this function! -> defer?
-        // 3. create IR & apply inference
-        todo!()
-    }
-}
-
-#[derive(Debug, PartialEq)]
-pub enum FunctionValuePrototype {
-    Thing(Type),
-    ThingOptional(Type),
-    Value(ValueType),
-    ValueOptional(ValueType),
-    ThingList(Type),
-    ThingListOptional(Type),
-    ValueList(ValueType),
-    ValueListOptional(ValueType),
+pub type StoredFunction = UncompiledFunction<DefinitionKey<'static>>;
+pub struct UncompiledFunction<IDType: FunctionIDTrait> {
+    pub(crate) function_id: IDType,
+    pub(crate) parsed: typeql::schema::definable::Function,
 }
```

Comment:
This no longer holds the IR.
All other information was held in the TypeQL function representation. 
The signature we needed in the IR package, so it was moved there and uses IR data-structures.
These names aren't great though.